Data/Vinyl/Index.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+
+module Data.Vinyl.Index where
+
+import Data.Vinyl.Core
+import Data.Vinyl.Functor
+
+-- | A Peano index provides a constructive proof that the type-level list @rs@
+-- contains the type @r@.
+data RIndex (rs :: [k]) (r :: k) where
+  Z :: RIndex (r ': rs) r
+  S :: RIndex rs r -> RIndex (s ': rs) r
+
+rlensIx
+  :: Functor g
+  => RIndex rs r
+  -> (f r     -> g (f r))
+  -> Rec f rs -> g (Rec f rs)
+rlensIx Z f (r:&rs) = fmap (:& rs) (f r)
+rlensIx (S ix) f (r:&rs) = fmap (r :&) (rlensIx ix f rs)
+
+rgetIx :: RIndex rs r -> Rec f rs -> f r
+rgetIx ix = getConst . rlensIx ix Const
+
+(!) :: Rec f rs -> RIndex rs r -> f r
+infixl 9 !
+(!) = flip rgetIx
+
+rputIx :: RIndex rs r -> f r -> Rec f rs -> Rec f rs
+rputIx ix r = getIdentity . rlensIx ix (\_ -> Identity r)
```

Comment:
You might want to use `coerce` instead of `getIdentity .` here for a minor potential speed-up.